src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -229,175 +225,162 @@ public void onJoinMessage(PlayerJoinEvent event) {
+         */
 
-        if (event.getAddress() == null) { // Address still null
-            unresolvedPlayerHostname.add(name);
+        if (from.getBlockX() == to.getBlockX()
+            && from.getBlockZ() == to.getBlockZ()
+            && from.getY() - to.getY() >= 0) {
             return;
-        } else {
-            unresolvedPlayerHostname.remove(name);
         }
 
-        if (!isAsyncPlayerPreLoginEventCalled || !settings.getProperty(PluginSettings.USE_ASYNC_PRE_LOGIN_EVENT)) {
-            try {
-                // Player.getAddress() can be null at this event, use event.getAddress()
-                runOnJoinChecks(JoiningPlayer.fromPlayerObject(player), event.getAddress().getHostAddress());
-            } catch (FailedVerificationException e) {
-                event.setKickMessage(messages.retrieveSingle(player, e.getReason(), e.getArgs()));
-                event.setResult(PlayerLoginEvent.Result.KICK_OTHER);
-            }
-        }
-    }
-
-    @EventHandler(priority = EventPriority.HIGHEST)
-    public void onPlayerQuit(PlayerQuitEvent event) {
         Player player = event.getPlayer();
+        if (!listenerService.shouldCancelEvent(player)) {
+            return;
+        }
 
-        if (settings.getProperty(RegistrationSettings.REMOVE_LEAVE_MESSAGE)) {
-            event.setQuitMessage(null);
-        } else if (settings.getProperty(RegistrationSettings.REMOVE_UNLOGGED_LEAVE_MESSAGE)) {
-            if (listenerService.shouldCancelEvent(event)) {
-                event.setQuitMessage(null);
-            }
+        if (!settings.getProperty(RestrictionSettings.ALLOW_UNAUTHED_MOVEMENT)) {
+            // "cancel" the event
+            event.setTo(event.getFrom());
+            return;
         }
 
-        if (antiBotService.wasPlayerKicked(player.getName())) {
+        if (settings.getProperty(RestrictionSettings.NO_TELEPORT)) {
             return;
         }
 
-        management.performQuit(player);
+        Location spawn = spawnLoader.getSpawnLocation(player);
+        if (spawn != null && spawn.getWorld() != null) {
+            if (!player.getWorld().equals(spawn.getWorld())) {
+                player.teleport(spawn);
+            } else if (spawn.distance(player.getLocation()) > settings.getProperty(ALLOWED_MOVEMENT_RADIUS)) {
+                player.teleport(spawn);
+            }
+        }
     }
 
     @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
-    public void onPlayerKick(PlayerKickEvent event) {
-        // Note #831: Especially for offline CraftBukkit, we need to catch players being kicked because of
-        // "logged in from another location" and to cancel their kick
-        if (settings.getProperty(RestrictionSettings.FORCE_SINGLE_SESSION)
-            && event.getReason().contains("You logged in from another location")) {
-            event.setCancelled(true);
+    public void onPlayerRespawn(PlayerRespawnEvent event) {
```

Comment:
This has no unit test :(

Reply:
It is so simple i have no idea how to test it in a way that makes sense xD

Reply:
It's a few tests:
1) No interaction with event if NO_TELEPORT is set to true
2) No interaction with event if listenerService.shouldCancelEvent is `false`
3) Finally, should set the spawn location to the event if we indeed get this far
4) Bonus: test that ensures everything is OK if the spawnLoader does not return a location ^^